src/app/core/data-services/data.service.ts

```diff
@@ -0,0 +1,44 @@
+import { OpaqueToken } from "@angular/core";
+import { Observable } from "rxjs";
+import { ObjectCacheService } from "../cache/object-cache.service";
+import { RequestCacheService } from "../cache/request-cache.service";
+import { CacheableObject } from "../cache/object-cache.reducer";
+import { ParamHash } from "../shared/param-hash";
+import { isNotEmpty } from "../../shared/empty.util";
+import { GenericConstructor } from "../shared/generic-constructor";
+
+export abstract class DataService<T extends CacheableObject> {
+  abstract serviceName: OpaqueToken;
+  protected abstract objectCache: ObjectCacheService;
+  protected abstract requestCache: RequestCacheService;
+
+  constructor(private modelType: GenericConstructor<T>) {
+
+  }
+
+  findAll(scopeID?: string): Observable<Array<T>> {
+    const key = new ParamHash(this.serviceName, 'findAll', scopeID).toString();
+    return this.requestCache.findAll(key, this.serviceName, scopeID)
+      //get an observable of the IDs from the RequestCache
+      .map(entry => entry.resourceUUIDs)
+      .flatMap((resourceUUIDs: Array<string>) => {
+        // use those IDs to fetch the actual objects from the ObjectCache
+        return this.objectCache.getList<T>(resourceUUIDs, this.modelType);
+      });
+  }
+
+  findById(id: string): Observable<T> {
+    const key = new ParamHash(this.serviceName, 'findById', id).toString();
+    return this.requestCache.findById(key, this.serviceName, id)
+      .map(entry => entry.resourceUUIDs)
+      .flatMap((resourceUUIDs: Array<string>) => {
+        if(isNotEmpty(resourceUUIDs)) {
+          return this.objectCache.get<T>(resourceUUIDs[0], this.modelType);
+        }
+        else {
+          return Observable.of(undefined);
```

Comment:
I'm not sure I agree. This is what `findById` is going to return, and that should be a single object. Either it exists and you return it, or it doesn't, and then returning `undefined` seems like a better idea than returning an empty array. You could argue for not returning anything at all (meaning that the observable doesn't fire a new event when it happens), but that would mean that if you have a UI element that's observing this object, and you clear it (e.g. because the user logged out, and no longer has access rights) the UI wouldn't receive an update, and keep on showing the object